pyro/poutine/continuation_poutine.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import absolute_import, division, print_function
+
+from .poutine import Messenger, Poutine
+
+
+class ContinuationMessenger(Messenger):
+    """
+    This Messenger sets the continuation field at each site,
+    and makes sure it gets called and unwound correctly.
+    Continuations give up (or simulate giving up) control of execution
+    by expanding along a dimension, yielding a coroutine,
+    or raising a special poutine.util.NonlocalExit exception.
+
+    It is currently used primarily for sequential and parallel enumeration.
+
+    Generalizes and replaces EscapeMessenger and EnumerateMessenger.
+    """
+    def __init__(self, escape_fn, cont_fn, first_available_dim):
```

Comment:
It seems like it would be cleaner to split this spit this into two classes: a base class `ContinuationMessenger` to handle `escape_fn,cont_fn` and a derived class `EnumerateMessenger` to handle `first_available_dim` and `next_available_dim`. It seems conflated to manage that dimension logic in e.g. `poutine.escape`. Also it will be easier to refactor dimension logic if we isolate that to the messenger that needs it, i.e. `EnumerateMessenger`.

Reply:
I kept it this way because this is mid-refactor; in the final version (if I ever figure out dependency tracking...) `ContinuationMessenger` will maintain an internal stack of `IndepMessenger`s to mark multiple invocations of the rest of the program independent.  These will be used for both `escape` and enumeration.

Basically, there's no reason for `ContinuationMessenger` to exist if it's not handling this logic.

Reply:
Ok, that's fine if it's mid-refactor. The main thing I'm trying to avoid is the leaking of specific details (like `first_available_dim`) into general interfaces. I found this to be awkward in e.g. `poutine.queue` which takes args `num_samples` and `max_tries`. I'd like to make a general effort to keep specific details of derived classes from leaking into base classes, so as to avoid all derived classes having to know about details of their siblings (and substitute oo with fp or your favorite abstraction mechanism).